integrations: Add support for Gitlab emoji events

Pritesh-30 opened the pull request:
Fixes #34221.

<!-- Describe your pull request here.-->
This is a replacement PR for #34303.

As suggested in the review, I have updated GitLab emoji reaction messages for comment targets to link directly to “a comment”, avoiding redundant target text as they are already present in the topic. 
**How changes were tested:**
Updated the relevant tests and verified that the full test suite passes locally.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
<img width="1129" height="144" alt="image" src="https://github.com/user-attachments/assets/25e16395-a3ad-4708-923c-1fe5d836eb3e" />
<img width="1138" height="150" alt="image" src="https://github.com/user-attachments/assets/d3aad963-15cd-4b56-b037-574085c24491" />
<img width="1121" height="137" alt="image" src="https://github.com/user-attachments/assets/55160bf1-e5f9-4ef7-8f18-2779f4dea199" />
<img width="1133" height="156" alt="image" src="https://github.com/user-attachments/assets/0b53bca1-dda9-4e53-b467-7163ee6c00e3" />
<img width="1132" height="162" alt="image" src="https://github.com/user-attachments/assets/0c92a5e2-d316-42bb-af63-f8e9cb645694" />



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

vishuverma-1 commented:
@Pritesh-30 is this issue resolved ?

Niloth-p commented on zerver/webhooks/gitlab/view.py:
```diff
@@ -474,6 +476,97 @@ def get_deployment_event_body(payload: WildValue, include_title: bool) -> str:
+    emoji = payload["object_attributes"]
+    awardable_type = emoji["awardable_type"].tame(check_string)
+    transformed_type = get_emoji_event_transformed_type(payload, awardable_type)
+    topic_title = get_emoji_event_topic_title(
```
I don't see why we need the title at all. `include_title` looks sufficient to me.

Niloth-p reviewed (commented):
Thanks for the updates! And the screenshot.

We should add more tests - there's just one test for `without_merge_request_title` right now, we should add tests with `with_custom_topic` too - maybe 1 each for snippet, MR and note.

All your edits are in a separate commit now, you need to use `git rebase` and edit the respective commits.

> If you're interested, you could add another commit at the end that updates similar duplicate suffixes in the rest of the integration, to be conditionally displayed if the topic does not have that info.

Are you planning to add such a commit?

Pritesh-30 commented:
> The Git commits are not correct. The changes don't belong to the commits they should. Please look into learning about the purpose of Git commits - their modularity. And about refactoring and layering Git commits. https://zulip.readthedocs.io/en/latest/git/index.html
> 
> 1. Make changes to Varun's PR by editing the respective commits.
> 2. Create a new commit for making the suffix optional in events that existed before this PR.
> 
> If you need more detail, refer back to my previous review comments about the commit structure.

@Niloth-p I have added the changes in the respective commits. Also made a new commit for events that existed before thi PR , to make suffix optional

timabbott commented on zerver/webhooks/gitlab/tests.py:
```diff
@@ -759,3 +759,47 @@ def test_deployment_canceled_event_message(self) -> None:
+
+    def test_emoji_award_in_snippet(self) -> None:
+        expected_topic_name = "sample / snippet #4831194 Sample Snippet"
+        expected_message = "Varun Kolanu added the emoji **thumbsup**."
```
(Ideally, we'd get the Unicode emoji, rather than a string name, since the string name databases may not match between Zulip and GitLab)

timabbott commented on zerver/webhooks/gitlab/fixtures/emoji_hook__award__design_note.json:
```diff
@@ -0,0 +1,97 @@
+    "object_attributes": {
+        "user_id": 26076637,
+        "created_at": "2025-04-05 08:18:27 UTC",
+        "id": 33657545,
```
Which it can't be, since `thumbsup` gets two different IDs.

So I think we probably will want to just issue `:heart:` type syntax in the message.

Niloth-p replied:
Do you mean we should use that syntax irrespective of whether there's a matchingly named emoji in Zulip? It gets rendered if it exists, else fine to render it with colons?

timabbott replied:
Right.

Pritesh-30 commented:
@timabbott @Niloth-p Sorry for the delay. I have updated the commits to use unicode for emojis. I have also updated the screenshots in PR description accordingly. As we had discussed earlier GitLab provides emoji like `thumbsup` and since in zulip we render it as `thumbs_up` so the message will be `:thumbsup:`

timabbott commented:
Merged, thanks @Pritesh-30 and @Niloth-p!
